File: backend/src/routes/activity.js
import express from 'express';
const router = express.Router();
import { auth } from '../middleware/auth.js';
import Activity from '../models/Activity.js';

router.get('/', auth, async (req, res) => {
  try {
    const activities = await Activity.find()
      .populate('userId', 'name email avatarUrl')
      .sort({ createdAt: -1 });

    res.json(activities.map(a => ({
      id: a._id.toString(),
      userId: a.userId._id.toString(),
      userName: a.userName,
      action: a.action,
      targetType: a.targetType,
      targetId: a.targetId,
      createdAt: a.createdAt.toISOString(),
    })));
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.get('/task/:taskId', auth, async (req, res) => {
  try {
    const activities = await Activity.find({
      targetType: 'task',
      targetId: req.params.taskId,
    })
      .populate('userId', 'name email avatarUrl')
      .sort({ createdAt: -1 });

    res.json(activities.map(a => ({
      id: a._id.toString(),
      userId: a.userId._id.toString(),
      userName: a.userName,
      action: a.action,
      targetType: a.targetType,
      targetId: a.targetId,
      createdAt: a.createdAt.toISOString(),
    })));
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.get('/user/:userId', auth, async (req, res) => {
  try {
    const activities = await Activity.find({ userId: req.params.userId })
      .populate('userId', 'name email avatarUrl')
      .sort({ createdAt: -1 });

    res.json(activities.map(a => ({
      id: a._id.toString(),
      userId: a.userId._id.toString(),
      userName: a.userName,
      action: a.action,
      targetType: a.targetType,
      targetId: a.targetId,
      createdAt: a.createdAt.toISOString(),
    })));
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

File: backend/app/routes/activity.py
from fastapi import APIRouter, Depends
from app.auth import get_current_user, transform_record
from app.database import get_admin_client

router = APIRouter()


@router.get("")
async def get_all_activities(current_user: dict = Depends(get_current_user)):
    supabase = get_admin_client()
    response = (
        supabase.table("activities")
        .select("*")
        .order("created_at", desc=True)
        .execute()
    )
    return [transform_record(a) for a in response.data]


@router.get("/task/{task_id}")
async def get_task_activities(
    task_id: str, current_user: dict = Depends(get_current_user)
):
    supabase = get_admin_client()
    response = (
        supabase.table("activities")
        .select("*")
        .eq("target_type", "task")
        .eq("target_id", task_id)
        .order("created_at", desc=True)
        .execute()
    )
    return [transform_record(a) for a in response.data]


@router.get("/user/{user_id}")
async def get_user_activities(
    user_id: str, current_user: dict = Depends(get_current_user)
):
    supabase = get_admin_client()
    response = (
        supabase.table("activities")
        .select("*")
        .eq("user_id", user_id)
        .order("created_at", desc=True)
        .execute()
    )
    return [transform_record(a) for a in response.data]
